src/core/experimental/rcptr.d

```diff
@@ -0,0 +1,208 @@
+// Written in the D programming language.
+/**
+This module provides a shared pointer implementation with memory management
+through reference counting.
+
+License: $(HTTP www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+Authors: Les De Ridder
+
+Source: $(DRUNTIMESRC core/experimental/rcptr.d)
+*/
+module core.experimental.rcptr;
+
+///
+unittest
+{
+    import core.stdc.stdlib : calloc;
+    auto allocInts = (size_t count) => cast(int*) calloc(count, int.sizeof);
+
+    __rcptr!int a; //default ctor
+    assert(a.get is null);
+
+    auto b = __rcptr!int(allocInts(10)); //ptr ctor
+
+    {
+        auto c = a; //copy ctor
+        assert(c.get == a.get);
+        assert(c.get != b.get);
+
+        c = b; //opAssign
+        assert(c.get != a.get);
+        assert(c.get == b.get);
+    }
+
+    assert(a.count is null);
+    assert(*b.count == 0);
+}
+
+/**
+A reference counted shared pointer that can be used to implement reference
+counted data structures.
+
+The reference count is automatically incremented/decremented on assignment,
+copy (construction), and destruction. When there are no more references to the
+pointer, the reference count is automatically deallocated and the referenced
+pointer is `free`d.
+
+Implementation: The internal implementation of `__rcptr` uses `malloc`/`free`.
+*/
+struct __rcptr(T)
+{
+    alias CounterType = int;
+
+    private T* ptr = null;
+    private shared(CounterType)* count = null;
+
+    /**
+    Creates a new `__rcptr` instance, tracking the provided pointer.
+
+    This implies that the ownership of the pointer is transferred to
+    `__rcptr`.
+
+    Params:
+         ptr = pointer to memory to be managed by `__rcptr`
+    */
+    this(inout(T*) ptr) inout
+    {
+        this.ptr = ptr;
+
+        if (ptr !is null)
+        {
+            import core.memory : pureCalloc;
+
+            // We use `calloc` so we don't have to manually initialise count/addRef
+            count = (() @trusted inout => cast(typeof(count)) pureCalloc(1, CounterType.sizeof))();
+        }
+    }
+
+    /**
+    Get the owned pointer.
+
+    Returns:
+        The pointer owned by this instance of `__rcptr`.
+    */
+    @system
+    inout(T*) get() inout
+    {
+        return ptr;
+    }
+
+    ~this()
+    {
+        delRef();
+    }
+
+    ///
+    void opAssign(__rcptr!T rhs)
```

Comment:
Now I'm really confused... you changed it to accept the arg by value, but then still ref-fiddling in the function?
All you need to do is swap the pointers. Taking the arg by value will increment the ref count, and then destruction of the local on return will decrement; you need to swap them before returning so it decrements the right object.

Sadly, this approach accepting the argument by-val does not optimise for the `rhs.count == count` case; it will do the ref fiddling in that case when it doesn't need to.

Reply:
Ah, my bad, I see what you're saying now. It does get a little messy because there is no swap function in druntime and we have to swap both `ptr` and `count`.

It indeed doesn't optimise for the `rhs.count == count` case though, but I'm not sure how much of an issue that is?

Reply:
Maybe we should have a `swap` function ;)